sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala

```diff
@@ -2128,38 +2128,67 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
+    )(withTempPath { path =>
+      val ds = sampledTestData.coalesce(1)
+      ds.write.text(path.getAbsolutePath)
+      val readback = spark.read.option("samplingRatio", 0.1).json(path.getCanonicalPath)
+
+      assert(readback.schema == new StructType().add("f1", LongType))
+    })
+  }
 
-      val ds = spark.read.option("samplingRatio", 0.1).json(path.getCanonicalPath)
-      assert(ds.schema == new StructType().add("f1", LongType))
-    }
+  test("SPARK-23849: usage of samplingRatio while parsing a dataset of strings") {
+    val ds = sampledTestData.coalesce(1)
+    val readback = spark.read.option("samplingRatio", 0.1).json(ds)
+
+    assert(readback.schema == new StructType().add("f1", LongType))
   }
 
-  test("SPARK-23849: usage of samplingRation while parsing of dataset of strings") {
-    val dstr = spark.sparkContext.parallelize(0 until 100, 1).map { i =>
-      val predefinedSample = Set[Int](2, 8, 15, 27, 30, 34, 35, 37, 44, 46,
-        57, 62, 68, 72)
-      if (predefinedSample.contains(i)) {
-        s"""{"f1":${i.toString}}""" + "\n"
-      } else {
-        s"""{"f1":${(i.toDouble + 0.1).toString}}""" + "\n"
-      }
-    }.toDS()
-    val ds = spark.read.option("samplingRatio", 0.1).json(dstr)
+  test("SPARK-23849: samplingRatio is out of the range (0, 1.0]") {
+    val ds = spark.range(0, 100, 1, 1).map(_.toString)
+
+    val errorMsg0 = intercept[IllegalArgumentException] {
+      spark.read.option("samplingRatio", -1).json(ds)
+    }.getMessage
+    assert(errorMsg0.contains("samplingRatio (-1.0) should be greater than 0"))
 
-    assert(ds.schema == new StructType().add("f1", LongType))
+    val errorMsg1 = intercept[IllegalArgumentException] {
+      spark.read.option("samplingRatio", 0).json(ds)
+    }.getMessage
+    assert(errorMsg1.contains("samplingRatio (0.0) should be greater than 0"))
+
+    val sampled = spark.read.option("samplingRatio", 10).json(ds)
+    assert(sampled.count() == ds.count())
+  }
+
+  test("SPARK-23849: sampling files for schema inferring in the multiLine mode") {
+    withTempDir { dir =>
+      Files.write(Paths.get(dir.getAbsolutePath, "0.json"),
+        """{"a":"a"}""".getBytes(StandardCharsets.UTF_8), StandardOpenOption.CREATE_NEW)
+      for (i <- 1 until 10) {
+        Files.write(Paths.get(dir.getAbsolutePath, s"$i.json"),
+          s"""{"a":$i}""".getBytes(StandardCharsets.UTF_8), StandardOpenOption.CREATE_NEW)
+      }
+      val files = (0 until 10).map { i =>
+        val hadoopConf = spark.sessionState.newHadoopConf()
+        val hdfsPath = new Path(Paths.get(dir.getAbsolutePath, s"$i.json").toString)
+        hdfsPath.getFileSystem(hadoopConf).getFileStatus(hdfsPath)
+      }
+      // The test uses the internal method because public API cannot guarantee order of files
+      // passed to the infer method. The order is changed between runs because the temporary
+      // folder has different path which leads to different order of file statuses returned
```

Comment:
> the temporary folder has different path which leads to different order of file statuses

Can we just read `dir` since we explicitly wrote the files above?

Reply:
Do you mean to pass a dir into the _infer()_ method instead of sequence of files?

Reply:
Maybe I missed something but if the order by different paths were problem, that's fixed by explicit file name above. Then, I thought we could just use `spark.read`?

Reply:
Ah, do you mean it's not guaranteed if we get the file statues in such order or not? Even if so, then we can gurantee `RDD[PortableDataStream]` is a single partition within  `MultiLineJsonDataSource.infer` .. @MaxGekk, it's okay if it's difficult to write a test. Manual test and updating PR description is a-okay.

Reply:
> Ah, do you mean it's not guaranteed if we get the file statues in such order or not, via, for example Hadoop API?

Yes, I am not sure we can guarantee that but frankly speaking I am not sure that sequence of files can guarantee that too. ;-)

> I took a look about this before and found it's designed to be in an alphabetical order although we shouldn't rely on this order if I remember this correctly.

Probably you are right. I ran the test 100 times by passing dir successfully but it says nothing. In another environment it can be flaky. 

> Even if so, then we can't gurantee `RDD[PortableDataStream]` is a single partition within `MultiLineJsonDataSource.infer`.

Right. From another hand it is pretty bad that we cannot control number of partitions and its sizes. Let's image sampled input is big enough and isn't evenly distributed across partitions. We will have the situation when most part of schema inferring job is performed by one task. Maybe it is better to repartition sampled RDD/Dataset before doing schema inferring. What do you think of it?

> it's okay if it's difficult to write a test. Manual test and updating PR description is a-okay.

Would you propose to delete the test?

Reply:
Yup. I think it's fine to delete.

Reply:
for repartition in schena inference, I dont feel strongly. lets monitor JIRAs and mailing list and see if there are some requests for it. it sounds too detailed control to me for now.